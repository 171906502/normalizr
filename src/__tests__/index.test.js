/* eslint-env jest */
import { normalize, schema } from '../';

describe('normalize', () => {
  [ 42, null, undefined, '42', () => {} ].forEach((input) => {
    it(`cannot normalize input that == ${input}`, () => {
      expect(() => normalize(input, new schema.Entity('test'))).toThrow();
    });
  });

  it('cannot normalize without a schema', () => {
    expect(() => normalize({})).toThrow();
  });

  it('normalizes entities', () => {
    const mySchema = new schema.Entity('tacos');

    expect(normalize([ { id: 1, type: 'foo' }, { id: 2, type: 'bar' } ], [ mySchema ])).toMatchSnapshot();
  });

  it('normalizes nested entities', () => {
    const user = new schema.Entity('users');
    const comment = new schema.Entity('comments', {
      user: user
    });
    const article = new schema.Entity('articles', {
      author: user,
      comments: [ comment ]
    });

    const input = {
      id: '123',
      title: 'A Great Article',
      author: {
        id: '8472',
        name: 'Paul'
      },
      body: 'This article is great.',
      comments: [
        {
          id: 'comment-123-4738',
          comment: 'I like it!',
          user: {
            id: '10293',
            name: 'Jane'
          }
        }
      ]
    };
    expect(normalize(input, article)).toMatchSnapshot();
  });

  it('does not modify the original input', () => {
    const user = new schema.Entity('users');
    const article = new schema.Entity('articles', { author: user });
    const input = Object.freeze({
      id: '123',
      title: 'A Great Article',
      author: Object.freeze({
        id: '8472',
        name: 'Paul'
      })
    });
    expect(() => normalize(input, article)).not.toThrow();
  });

  it('ignores null values', () => {
    const myEntity = new schema.Entity('myentities');
    expect(normalize([ null ], [ myEntity ])).toMatchSnapshot();
    expect(normalize([ undefined ], [ myEntity ])).toMatchSnapshot();
    expect(normalize([ false ], [ myEntity ])).toMatchSnapshot();
  });

  it('can use fully custom entity classes', () => {
    class MyEntity extends schema.Entity {
      schema = {
        children: [ new schema.Entity('children') ]
      };

      getId(entity, parent, key) {
        return entity.uuid;
      }

      normalize(input, parent, key, visit, addEntity) {
        const entity = { ...input };
        Object.keys(this.schema).forEach((key) => {
          const schema = this.schema[key];
          entity[key] = visit(input[key], input, key, schema, addEntity);
        });
        addEntity(this, entity, parent, key);
        return {
          uuid: this.getId(entity),
          schema: this.getKey(input, parent, key)
        };
      }
    }

    const mySchema = new MyEntity('food');
    expect(normalize({
      uuid: '1234',
      name: 'tacos',
      children: [
        { id: 4, name: 'lettuce' }
      ]
    }, mySchema)).toMatchSnapshot();
  });

  it('uses the non-normalized input when getting the key for an entity', () => {
    const userEntity = new schema.Entity('users');
    const inferKeyFn = jest.fn((nonNormalized, parent, key) => nonNormalized.user.type);
    const recommendation = new schema.Entity(inferKeyFn, { user: userEntity });
    expect(normalize({
      id: '123', user: { id: '456', type: 'user_recommendation' }
    }, recommendation)).toMatchSnapshot();
    expect(inferKeyFn.mock.calls).toMatchSnapshot();
  });

  it('uses the non-normalized input when getting the ID for an entity', () => {
    const userEntity = new schema.Entity('users');
    const idAttributeFn = jest.fn((nonNormalized, parent, key) => nonNormalized.user.id);
    const recommendation = new schema.Entity('recommendations', { user: userEntity }, {
      idAttribute: idAttributeFn
    });
    expect(normalize({ user: { id: '456' } }, recommendation)).toMatchSnapshot();
    expect(idAttributeFn.mock.calls).toMatchSnapshot();
  });

  it('passes over pre-normalized values', () => {
    const userEntity = new schema.Entity('users');
    const articleEntity = new schema.Entity('articles', { author: userEntity });

    expect(normalize({ id: '123', title: 'normalizr is great!', author: 1 }, articleEntity)).toMatchSnapshot();
  });
});
