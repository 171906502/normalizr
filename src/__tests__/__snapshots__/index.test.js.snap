exports[`normalize can use fully custom entity classes 1`] = `
Object {
  "entities": Object {
    "children": Object {
      "4": Object {
        "id": 4,
        "name": "lettuce",
      },
    },
    "food": Object {
      "1234": Object {
        "children": Array [
          4,
        ],
        "name": "tacos",
        "uuid": "1234",
      },
    },
  },
  "result": Object {
    "schema": "food",
    "uuid": "1234",
  },
}
`;

exports[`normalize ignores null values 1`] = `
Object {
  "entities": Object {},
  "result": Array [
    null,
  ],
}
`;

exports[`normalize ignores null values 2`] = `
Object {
  "entities": Object {},
  "result": Array [
    undefined,
  ],
}
`;

exports[`normalize ignores null values 3`] = `
Object {
  "entities": Object {},
  "result": Array [
    false,
  ],
}
`;

exports[`normalize normalizes entities 1`] = `
Object {
  "entities": Object {
    "tacos": Object {
      "1": Object {
        "id": 1,
        "type": "foo",
      },
      "2": Object {
        "id": 2,
        "type": "bar",
      },
    },
  },
  "result": Array [
    1,
    2,
  ],
}
`;

exports[`normalize normalizes nested entities 1`] = `
Object {
  "entities": Object {
    "articles": Object {
      "123": Object {
        "author": "8472",
        "body": "This article is great.",
        "comments": Array [
          "comment-123-4738",
        ],
        "id": "123",
        "title": "A Great Article",
      },
    },
    "comments": Object {
      "comment-123-4738": Object {
        "comment": "I like it!",
        "id": "comment-123-4738",
        "user": "10293",
      },
    },
    "users": Object {
      "10293": Object {
        "id": "10293",
        "name": "Jane",
      },
      "8472": Object {
        "id": "8472",
        "name": "Paul",
      },
    },
  },
  "result": "123",
}
`;
